src/mlpack/methods/kmeans/naive_kmeans_impl.hpp

```diff
@@ -38,28 +40,52 @@ double NaiveKMeans<MetricType, MatType>::Iterate(const arma::mat& centroids,
   counts.zeros(centroids.n_cols);
 
   // Find the closest centroid to each point and update the new centroids.
-  for (size_t i = 0; i < dataset.n_cols; i++)
+  // Computed in parallel over the complete dataset
+  #pragma omp parallel
   {
-    // Find the closest centroid to this point.
-    double minDistance = std::numeric_limits<double>::infinity();
-    size_t closestCluster = centroids.n_cols; // Invalid value.
+    // The current state of the K-means is private for each thread
+    arma::mat localCentroids(centroids.n_rows, centroids.n_cols,
+                             arma::fill::zeros);
+    arma::Col<size_t> localCounts(centroids.n_cols, arma::fill::zeros);
 
-    for (size_t j = 0; j < centroids.n_cols; j++)
+    #ifdef _WIN32
+    // Tiny workaround: Visual Studio only implements OpenMP 2.0, which doesn't
+    // support unsigned loop variables. If we're building for Visual Studio, use
+    // the intmax_t type instead.
+    #pragma omp for
+    for (intmax_t i = 0; i < (intmax_t) dataset.n_cols; ++i)
```

Comment:
Hmm, I think this pattern appears a bunch of places now.  We might want to add to `prereqs.hpp`:

```
// This can be removed with Visual Studio supports an OpenMP version with unsigned loop variables.
#ifdef _WIN32
  #define omp_size_t intmax_t
#else
  #define omp_size_t size_t
#endif
```

Then for the loops,

```
#pragma omp for
for (omp_size_t i = 0; i < (omp_size_t) dataset.n_cols; ++i)
```

But maybe that is out of scope for this PR.  I just wanted to point it out as an idea.